java-squid/src/main/java/org/sonar/java/cfg/CFG.java

```diff
@@ -0,0 +1,750 @@
+  }
+
+  private void buildDoWhileStatement(DoWhileStatementTree tree) {
+    DoWhileStatementTree s = tree;
+    Block falseBranch = currentBlock;
+    Block loopback = createBlock();
+    // process condition
+    currentBlock = createBranch(s, loopback, falseBranch);
+    buildCondition(s.condition(), loopback, falseBranch);
+    // process body
+    currentBlock = createBlock(currentBlock);
+    continueTargets.addLast(loopback);
+    breakTargets.addLast(falseBranch);
+    build(s.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    loopback.successors.add(currentBlock);
+    currentBlock = createBlock(currentBlock);
+  }
+
+  private void buildForEachStatement(ForEachStatement tree) {
+    ForEachStatement s = tree;
+    // TODO(npe) One solution is to create a forstatement node depending on type of expression (iterable or array) and build CFG from
+    // it.
+    Block afterLoop = currentBlock;
+    Block loopback = createBlock();
+    continueTargets.addLast(loopback);
+    breakTargets.addLast(afterLoop);
+    build(s.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    build(s.variable());
+    build(s.expression());
+  }
+
+  private void buildForStatement(ForStatementTree tree) {
+    ForStatementTree s = tree;
+    Block falseBranch = currentBlock;
+    // process step
+    currentBlock = createBlock();
+    Block updateBlock = currentBlock;
+    for (StatementTree updateTree : Lists.reverse(s.update())) {
+      build(updateTree);
+    }
+    continueTargets.addLast(currentBlock);
+    // process body
+    currentBlock = createBlock(currentBlock);
+    breakTargets.addLast(falseBranch);
+    build(s.statement());
+    breakTargets.removeLast();
+    continueTargets.removeLast();
+    Block body = currentBlock;
+    // process condition
+    ExpressionTree condition = s.condition();
+    if (condition != null) {
+      currentBlock = createBranch(s, body, falseBranch);
+      buildCondition(condition, body, falseBranch);
+    } else {
+      currentBlock = createUnconditionalJump(s, body);
+    }
+    updateBlock.successors.add(currentBlock);
+    // process init
+    currentBlock = createBlock(currentBlock);
+    for (StatementTree init : Lists.reverse(s.initializer())) {
+      build(init);
+    }
+  }
+
+  private void buildTryStatement(TryStatementTree tree) {
+    // FIXME only path with no failure constructed for now, (not taking try with resources into consideration).
+    TryStatementTree tryStatementTree = tree;
+    currentBlock = createBlock(currentBlock);
+    BlockTree finallyBlock = tryStatementTree.finallyBlock();
+    if (finallyBlock != null) {
+      build(finallyBlock);
+    }
+    currentBlock = createBlock(currentBlock);
+    build(tryStatementTree.block());
+  }
+
+  private void buildThrowStatement(ThrowStatementTree tree) {
+    // FIXME this won't work if it is intended to be caught by a try statement.
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1134)